doc/conf.py

```diff
@@ -939,3 +984,8 @@ def setup(app):
     linkcheck_request_headers = {
         "https://github.com/": {"Authorization": f"token {github_token}"},
     }
+
+# Write the pages prior to any sphinx event
+generate_index_rst()
+generate_min_dependency_table()
+generate_min_dependency_substitutions()
```

Comment:
It's not possible to use `app.connect` with an early sphinx event and priority setting instead of putting those functions in the `doc/conf.py`'s `__main__`?

`config-inited` should do, no?

https://www.sphinx-doc.org/en/master/extdev/appapi.html#events

Reply:
Also why `generate_min_dependency_table` and `generate_min_dependency_substitutions` need to be executed earlier in the build process than previously? I don't see there dependency on `parsed_version.is_devrelease` for those two.

Reply:
Since these generation does not depend on `app` I think we can just call them directly in `conf.py`? Personally I would prefer not to play with these events when it is not necessary. For instance, when playing around I tried to generate file with `.. autosummary::` and register the function to `config-inited` with very high priority, but they failed to be detected and I could not figure out why.

Maybe the comment should be changed, or do you still prefer registering to events?

Reply:
It's just that I find it weird to have a Python module named `conf.py` that is usually purely declarative have a side effect.

I am not a sphinx expert, I would have thought that `conf.py` is meant to host declarative configuration for sphinx commands and those commands should trigger side effect based on the updated configuration.

> For instance, when playing around I tried to generate file with .. autosummary:: and register the function to config-inited with very high priority, but they failed to be detected and I could not figure out why.

If the event system makes it overly complicated to get the behavior you want, then ok to keep it this way, but I have the feeling it's a slight abuse of the sphinx conventions.

Reply:
From [sphinx documentation](https://www.sphinx-doc.org/en/master/usage/configuration.html):

> The configuration file is executed as Python code at build time (using [importlib.import_module()](https://docs.python.org/3/library/importlib.html#importlib.import_module), and with the current directory set to its containing directory), and therefore can execute arbitrarily complex code. Sphinx then reads simple names from the file’s namespace as its configuration.

Not sure if by saying *can execute arbitrarily complex code*, it means that `conf.py` is allowed to trigger side effect or not. I looked around and found that `pandas` is also writing files in `conf.py` so perhaps it is okay.

https://github.com/pandas-dev/pandas/blob/fc37c83c2deb59253ae2d10ca631b6a1242cfdcb/doc/source/conf.py#L119-L127

Reply:
Alright then.